src/analysis/interpolation/qgstininterpolator.cpp

```diff
@@ -294,14 +294,21 @@ int QgsTinInterpolator::insertData( const QgsFeature &f, QgsInterpolator::ValueS
               switch ( source )
               {
                 case ValueAttribute:
-                  point.setZ( attributeValue );
+                  if ( point.is3D() )
+                    point.setZ( attributeValue );
+                  else
+                    point.addZValue( attributeValue );
                   break;
 
-                case ValueZ:
+                case ValueM:
+                  if ( point.is3D() )
+                    point.setZ( point.m() );
```

Comment:
Substituting m with z ison purpose?

Reply:
it was here before the fix, so I kept the logic...